### PR TITLE
Custom JSX Parser

### PR DESCRIPTION
**Pull request by developit**
This PR replaces `htm`'s DOM-based parsing with a custom parser implementing the complete JSX syntax.

The result is a fair bit smaller (~600b), produces higher-quality intermediary code, and likely consumes/retains a lot less memory since no objects are created.

/cc @jviide sorry for taking so long to get this PR'd!

### CODE REVIEW

File: src/index.mjs [hunk 1]
```diff
@@ -14,9 +14,9 @@
 const CACHE = {};
 
 export default function html(statics) {
-	const key = statics;
-
+	const key = statics.length + statics;
```
Comment (jviide) [inline src/index.mjs]: This has a key collision in cases such as `<a>${foo},</a>` and `<a>,${foo}</a>`.

File: test/index.test.mjs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+		expect(html`<a ...${{ foo: 'bar' }} b c />`).toEqual({ tag: 'a', props: { b: true, c: true, foo: 'bar' }, children: [] });
+	});
+
+	test('closing tag', () => {
+		expect(html`<a></a>`).toEqual({ tag: 'a', props: null, children: [] });
+		expect(html`<a b></a>`).toEqual({ tag: 'a', props: { b: true }, children: [] });
+	});
+
+	test('auto-closing tag', () => {
+		expect(html`<a><//>`).toEqual({ tag: 'a', props: null, children: [] });
```
Comment (dy) [inline test/index.test.mjs]: Not related to the PR - but what was the motivation for double-slash closing tag? JSX introduces `</>` for the purpose of closing fragments, and tbh I had that expectation here too. Are there some collisions?

Comment (developit) [reply test/index.test.mjs]: The original reason was that it was easy to parse - end tag names were ignored in the HTML parser version and are ignored in the custom parser version too. Really you could give the tag any name, like `<a></>`.

Comment (dy) [reply test/index.test.mjs]: Does the new parser ignore end tags due to the same reason?

Comment (developit) [reply test/index.test.mjs]: just to save bytes. It doesn't validate the opening and closing tags match.

**Pull request merged**